state-reconstruct-fetcher/src/types/v3.rs

```diff
@@ -90,60 +95,126 @@ impl TryFrom<&abi::Token> for V3 {
+        blobs.append(&mut blob_bytes);
+        pointer += PUBDATA_COMMITMENT_SIZE;
     }
-}
 
-fn parse_pubdata_from_calldata(
-    bytes: &[u8],
-    pointer: &mut usize,
-) -> Result<Vec<L2ToL1Pubdata>, ParseError> {
-    let mut l2_to_l1_pubdata = Vec::new();
+    l = blobs.len();
+    while l > 0 && blobs[l - 1] == 0u8 {
+        l -= 1;
+    }
 
-    // Skip over logs and messages.
-    let num_of_l1_to_l2_logs = u32::from_be_bytes(read_next_n_bytes(bytes, pointer));
-    *pointer += L2_TO_L1_LOG_SERIALIZE_SIZE * num_of_l1_to_l2_logs as usize;
+    let blobs_view = &blobs[..l];
+    parse_resolved_pubdata(blobs_view)
+}
 
-    let num_of_messages = u32::from_be_bytes(read_next_n_bytes(bytes, pointer));
-    for _ in 0..num_of_messages {
-        let current_message_len = u32::from_be_bytes(read_next_n_bytes(bytes, pointer));
-        *pointer += current_message_len as usize;
+async fn get_blob(kzg_commitment: &[u8], client: &BlobHttpClient) -> Result<Vec<u8>, ParseError> {
+    let url = client.format_url(kzg_commitment);
+    for attempt in 1..=MAX_RETRIES {
+        match client.retrieve_url(&url).await {
+            Ok(response) => match response.text().await {
+                Ok(text) => match get_blob_data(&text) {
+                    Ok(data) => {
+                        let plain = if let Some(p) = data.strip_prefix("0x") {
+                            p
+                        } else {
+                            &data
+                        };
+                        return hex::decode(plain).map_err(|e| {
+                            ParseError::BlobFormatError(plain.to_string(), e.to_string())
+                        });
+                    }
+                    Err(e) => {
+                        tracing::error!("failed parsing response of {url}");
+                        return Err(e);
+                    }
+                },
+                Err(e) => {
+                    tracing::error!("attempt {}: {} failed: {:?}", attempt, url, e);
+                    sleep(Duration::from_secs(FAILED_FETCH_RETRY_INTERVAL_S)).await;
+                }
+            },
+            Err(e) => {
+                tracing::error!("attempt {}: GET {} failed: {:?}", attempt, url, e);
+                sleep(Duration::from_secs(FAILED_FETCH_RETRY_INTERVAL_S)).await;
+            }
+        }
     }
+    Err(ParseError::BlobStorageError(url))
+}
 
-    // Parse published bytecodes.
-    let num_of_bytecodes = u32::from_be_bytes(read_next_n_bytes(bytes, pointer));
-    for _ in 0..num_of_bytecodes {
-        let current_bytecode_len = u32::from_be_bytes(read_next_n_bytes(bytes, pointer)) as usize;
-        let bytecode = bytes[*pointer..*pointer + current_bytecode_len].to_vec();
-        *pointer += current_bytecode_len;
-        l2_to_l1_pubdata.push(L2ToL1Pubdata::PublishedBytecode(bytecode))
+fn get_blob_data(json_str: &str) -> Result<String, ParseError> {
+    if let Ok(v) = serde_json::from_str(json_str) {
```

Comment:
```suggestion
    let Ok(v) = serde_json::from_str(json_str) else {
        return Err(ParseError::BlobFormatError(
            json_str.to_string(),
            "not JSON".to_string(),
        ));
    };
```
`Option/Result`-chains, especially where we just return an `Err` in the other case, are better expressed as `let else` statements.

Reply:
OK